# **수업에 들어가기 전 ...**
['AI 객체 탐지 모델로 작업자의 위험 행동을 감지한다'](https://www.youtube.com/watch?v=dfXlxFmqrNk)

# **강의목표**

1. 객체탐지의 핵심 개념을 이해하고, 단순 이미지 분류와의 차이점을 설명할 수 있다.
2. Ultralytics를 사용하여 사전 학습된 YOLO 모델을 로드하고, 주어진 이미지에서 객체를 탐지하는 코드를 작성 및 실행할 수 있다.
3. 탐지된 결과의 의미를 해석하고, 이를 응용하여 실생활 속 문제를 어떻게 해결할 수 있을 지 아이디어를 제시할 수 있다.

# 🤖 **1. 객체 탐지 (Object Detection)**

컴퓨터 비전 기술 중 하나로, 이미지나 영상 속에서 **<U>특정 객체의 종류(what)와 위치(where)를 찾아내는 기술</U>**, 단순히 이미지를 분류하는 것을 넘어, 객체 주변에 경계 상자(bounding box)를 그리고 어떤 객체인지 레이블(label)을 붙여주는 것이 핵심!

## 👉 1-1. 이미지 분류와의 차이점

- **Classification(분류):** 여러 물체에 대해 어떤 물체인지 <U>**클래스를 분류**</U>하는 문제
- **Localization(위치 파악):** 물체가 어디 있는지 박스를 통해 <U>**위치 정보**</U>를 나타내는 문제
- **Object Detection(객체탐지):** Classification과 Localization을 <U>**동시에 진행**</U>하는 것

<img src="https://drive.google.com/uc?export=view&id=1sV3qSNESkRW-6pYL4BcrmqBTGWqbzVqR">

# 🤖 **2. YOLO (You Only Look Once, 단 한번만 본다)**

이미지나 영상 속에서 물체(객체)의 위치와 종류를 **<U>'단 한 번만 보고'</U>**
한 번에 찾아내는 딥러닝 기반의 객체 탐지(Object Detection) 모델


## 👉 2-1. YOLO 버전 별 주요 개선사항 및 특징



| 연도         | 버전       | 주요 개선 사항 및 특징                                                                                                                                                                                                      |
|--------------|------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| 2016         | YOLO (v1)  | - 워싱턴 대학교 조셉 레드몬, 알리 파르하디가 개발<br>- 단일 단계(single-stage) 객체 탐지 기법 도입<br>- 이미지를 S×S 격자로 나누어 위치+클래스를 동시에 예측<br>- 7×7 그리드, 각 셀에서 2개의 바운딩 박스와 클래스 확률 출력 |
| 2017         | YOLOv2     | - “YOLO9000: Better, Faster, Stronger” 논문<br>- 앵커 박스 개념 도입(Faster R-CNN에서 차용)<br>- 차원 클러스터링(Anchor Box 크기 최적화)<br>- 다중 스케일(multiscale) 학습과 일괄 정규화(BN) 적용                                 |
| 2018         | YOLOv3     | - Darknet-53 백본 사용(잔차 블록+스킵 커넥션 구조)<br>- 3단계(3-scale) 예측으로 작은 물체 인식 성능 개선<br>- 시그모이드 함수를 통한 바운딩 박스 좌표 예측<br>- 공간 피라미드 풀링(SPP) 아이디어 일부 반영                          |
| 2020년 4월   | YOLOv4     | - CSPDarknet53 백본 도입 및 Mosaic 데이터 증강<br>- BoF(Freebies), BoS(Specials) 기법 등 다양한 최적화 기법 집약<br>- SPP 추가, Weighted-Residual-Connection(WRC) 적용<br>- 빠른 추론 속도와 높은 정확도 달성                |
| 2020년 6월   | YOLOv5     | - Ultralytics에서 PyTorch 기반으로 개발<br>- 자동 앵커(Auto Anchor), 모자이크 증강 등 다양한 편의 기능<br>- 하이퍼파라미터 최적화 및 통합 실험 추적 제공<br>- 모델 크기가 작아지고, v4 대비 약 10% 이상 성능 향상                     |
| 2022년 7월   | YOLOv7     | - 기존 YOLO 계열 대비 학습 과정 최적화 및 연산량 감소<br>- E-ELAN(Extended ELAN) 구조를 활용하여 성능 강화<br>- trainable bag of freebies 등 학습 전략 다수 제안<br>- 실시간 객체 탐지 분야에서 높은 정확도 및 속도 제공         |
| 2022년 9월   | YOLOv6     | - Meituan(메이투안)에서 개발<br>- 앵커 기반 + 앵커 프리 방식 모두 지원<br>- Quantization, Knowledge Distillation 기법 도입<br>- 효율적 구조 재설계를 통해 속도·정확도 동시 개선                                           |
| 2023년 1월   | YOLOv8     | - Ultralytics에서 새롭게 공개한 통합 프레임워크<br>- 앵커 프리(Anchor-free) 기법 및 디커플 헤드(Decoupled Head) 적용<br>- 객체 감지, 인스턴스 세분화 등 다양한 시각지능 모델링에 사용<br>- 이전 버전 대비 가볍고 정확도 개선         |
| 2024년 2월   | YOLOv9     | - 기존 네트워크에서 정보 손실 문제를 해결하기 위해 PGI(Preserved Gradient Integration) 적용<br>- GELAN(Generalized ELAN) 신경망 구조로 업그레이드<br>- 연산 효율과 정확도를 동시에 향상                                           |
| 2024년 5월   | YOLOv10    | - 칭화대 연구진이 Ultralytics Python 패키지를 활용해 개발<br>- 비최대 억제(NMS) 제거한 엔드투엔드 감지 헤드 도입<br>- 고난도 환경에서의 물체 분할 및 인식 정확도 개선                                                               |
| 2024년 9월   | YOLOv11    | - 물체 감지, 세분화, 포즈 추정, 추적, 분류 등 다양한 작업에서 최첨단(SOTA) 성능 제공<br>- Ultralytics YOLO 계열의 최신 버전<br>- 확장된 멀티태스킹 능력으로 산업·연구 분야에서 활용도 증대   

- **주요 개선사항 요약**  
  > - **YOLO ~ YOLOv3 (기본 규칙과 뼈대)**: 격자 기반 단일 단계 예측 구조를 유지하면서 앵커 박스, 멀티 스케일 예측 등 주요 기능 추가  
  > - **YOLOv4 ~ YOLOv7 (만들어진 뼈대에 살과 근육을 붙임)**: 백본 네트워크 개선(CSPDarknet53, E-ELAN 등), 다양한 데이터 증강(Mosaic 등) 및 학습 최적화 기법 도입  
  > - **YOLOv8 ~ (기존 방식에서 벗어나 더 효율적인 방식)**: 앵커 프리, 디커플 헤드, 엔드투엔드 감지 헤드 등 새로운 접근으로 정확도·속도 동시 향상  




****
- **YOLO 모델 버전/크기 별 성능(정확도)과 처리 속도의 관계**
<img src="https://drive.google.com/uc?export=view&id=15GMBhUvu8v1TfikOj152Uw6ENlq7Lf9B">


> - `Latency T4 TensorRT10 FP16` : Nvidia T4 GPU에서 TensorRT 10 최적화와 FP16 연산 방식을 사용하여, **모델이 얼마나 빠르게 예측을 수행하는지**를 측정한 값 => 왼쪽으로 갈수록 속도가 더 빠른 모델
> - `COCO mAP` : COCO 데이터셋을 기준으로, 모델이 **객체를 얼마나 정확하게 감지하는지**를 평가하는 대표적인 성능 지표 => 위쪽으로 갈수록 정확도가 높은 모델

<br>
<br>

| 구분 | 작은 모델 (n, s) | 큰 모델 (l, x) |
| :--- | :--- | :--- |
| **비유** | 작은 동물의 뇌 | 인간의 뇌 |
| **파라미터 수** | 적음 (수백만 개) | 많음 (수천만 개 이상) |
| **장점** | ✅ 매우 빠름, 가벼움, 저사양 장비 가능 | ✅ 매우 정확함, 복잡한 문제 해결 |
| **단점**| ❌ 정확도 낮음, 단순한 패턴만 학습 | ❌ 매우 느림, 무거움, 고사양 장비 필수 |
| **주요 용도**| 실시간 영상 분석, 모바일 기기 | 고정밀 분석, 의료/연구 분야 |

## 👉 2-2. YOLO 이론

### 1) YOLO 모델 이전의 객체 탐지

<img src="https://drive.google.com/uc?export=view&id=1bDzNBBX-eIUrXFIdi7vlTW5He_kL_hUp">


- R-CNN은 **Regions with CNN features**의 약자로 이미지 속에서 객체가 있을 만한 <U>**영역(Regions)**</U>을 먼저 찾고, 그 영역들을 <U>**CNN(딥러닝)**</U>을 이용해 분석하는 방식

- 작동 단계
  - 1. 후보 영역 제안 (Region Proposal) : Selective Search, Sliding window
  - 2. 이미지 변형 (후보 영역 이미지를 CNN 모델 입력사이즈에 맞게 변형
  - 3. CNN 특징 추출
  - 4. 분류 및 위치 보정

- 장점 : 딥러닝을 적용하여 월등히 높은 정확도를 달성함
- 단점 : 치명적으로 느린 속도, 복잡한 구조(학습과정이 복잡하고 비효율적임)

### 2) YOLO 모델 등장

YOLO는 **You Only Look Once**의 약자로, '후보 영역 제안'과 '분류'를 나누지 않고, <U>이미지 전체</U>를 단 한번만 보고 객체의 <U>위치와 종류를 동시에 예측</U>하는 <U>**1-Stage Detector**</U> 방식

<img src="https://drive.google.com/uc?export=view&id=1IAbW4L9bTbPhgb8DuuKsaKqTTAwYoEeh">


- 작동 단계
  - 1. 이미지 그리드 분할 : S*S 형태의 격자(Grid)로 나눔
  - 2. 단일 네트워크로 동시 예측 : 객체의 Bounding Box와 해당 객체가 무엇일지 동시에 예측
  - 3. 최종 결과 필터링

- 장점 : 압도적으로 빠른속도, 이미지 전체를 보기 때문에 맥락을 이해하여 배경 오류가 적음, 간결한 구조
- 단점 : 상대적으로 낮은 정확도


### 3) YOLO 모델의 원리

<img src="https://drive.google.com/uc?export=view&id=17cvLgkxJlDtulQ28wo8kjOXS-N5r5duS">


> **1단계 - 그리드 분할 :**  입력된 이미지를 S x S 크기의 격자(Grid)로 나눕니다.
>
> **2단계 - 동시 예측 :**  각 격자 셀이 위치와 클래스 정보를 동시에 예측합니다.
> - **Bounding Box 예측:** 객체의 위치(x,y), 크기(w,h), 객체 존재 신뢰도(confidence)를 예측합니다.
> - **클래스 확률 예측:** 해당 셀에 객체가 있을 경우, 어떤 종류(class)일지에 대한 확률을 예측합니다.
>
> **3단계 - 결과 종합 :**  모든 셀의 예측 결과를 하나로 모아 수많은 후보 상자를 생성합니다.
>
> **4단계 - 후처리 (Post-processing) :**  신뢰도가 낮은 상자를 제거하고, 중복된 상자들 중 가장 점수가 높은 하나만 남깁니다 (NMS, Non-Maximum Suppression).

### 4) YOLO 아키텍처

<img src="https://drive.google.com/uc?export=view&id=1AyJWwLdBXH-YOfqXLbX-ot0LFMRrXMgT">

> **초기 레이어 :** 이미지 압축 및 기본 특징 추출 (넓고 얕은 정보)
>
> **중간 레이어 :** 특징의 고도화 (좁고 깊은 지식)
>
> **최종 레이어 :** 최종 특징 맵 도출 (이미지의 모든 핵심 정보 압축)

## 👉 2-3. YOLO 모델 평가지표

### YOLO 모델 평가지표 : 모델의 성능을 다양한 관점에서 평가

| 지표                           | 설명                                                                                                                       | 특징 / 활용                                                         |
|--------------------------------|----------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------|
| **IoU (Intersection over Union)**  | - 예측 박스와 실제 박스의 겹치는 영역 비율<br>- 0~1 범위, 1은 완벽 일치                                                    | - 객체 탐지의 기본 기준<br>- mAP, AP50/75 계산 시 임계값 설정에 사용됨  |
| **Precision(정밀도) / Recall(재현율) / F1**    | - **Precision:** 예측한 객체 중 실제 객체 비율<br>- **Recall:** 실제 객체 중 예측 성공 비율<br>- **F1:** 둘의 조화 평균       | - 세밀한 모델 성능 분석에 적합<br>- 클래스 불균형 상황에서 유용         |
| **mAP (mean Average Precision)**  | - 여러 클래스의 AP(Average Precision) 평균값<br>- IoU 임계값(예: 0.5, 0.75 등)에 따라 평가                                    | - 모델 전반의 정확도를 종합적으로 평가<br>- 객체 탐지 분야의 대표 지표     |
| **AP50 / AP75**                | - **AP50:** IoU 0.5 기준 AP<br>- **AP75:** IoU 0.75 기준, 더 엄격한 평가                                                     | - IoU 기준에 따라 모델의 예측 정확도 세분화<br>- 프로토타입 성능 비교에 활용 |
| **FPS (Frames Per Second)**    | - 초당 처리 가능한 이미지(프레임) 수                                                                                        | - 실시간 응용(자율주행, CCTV 등)에서 중요<br>- 속도 평가의 주요 지표       |
| **Inference Time / Latency**   | - 단일 이미지 처리 시간 (밀리초 단위)                                                                                        | - 시스템 반응 속도 파악<br>- 배포 환경 및 하드웨어 최적화 판단에 활용       |


> - 전이학습 후 모델의 정확도를 전반적으로 확인하려면 **mAP**가 가장 대표적 (전체적인 정확도)
> - 특정 IoU 임계값에서의 성능을 더 자세히 보려면 **AP50** 또는 **AP75** 활용 (객체의 위치 탐색의 정밀도)
> - 클래스 불균형 여부나 세밀한 오탐/미탐을 파악하려면 **Precision/Recall/F1** 지표 병행 (각 클래스에 대한 성능 세밀 분석)
> - 공간적 예측의 정확도는 **IoU**로 판단(예측 박스), 속도는 **FPS/Inference Time**을 확인(실시간 처리 능력)

# 🤖 **3. Ultralytics**

컴퓨터 비전 분야의 소프트웨어를 만드는 회사 (YOLO 핵심 개발사)

- Ultralytics Yolo 가이드 링크 : https://docs.ultralytics.com/
- 모델 설명 링크 : https://docs.ultralytics.com/models/


# 🤖 **4. 실습) 음식이미지 분류 모델 만들기**
<img src="https://drive.google.com/uc?export=view&id=1fR2C-AWlw76zR_2lDWJhZz2x8fry1Xf7">

## 👉 4-1. 구글 드라이브 접근 (이미지 파일 로드 위함)

In [85]:
#코랩에서 구글 드라이브 기능 사용
from google.colab import drive

drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [86]:
#현재 위치(경로) 확인
%pwd

'/content/drive/My Drive/AI/컴퓨터비전'

In [87]:
#현재 폴더(드라이브-Ultralytics)로 이동
%cd '/content/drive/MyDrive/AI/컴퓨터비전'

/content/drive/MyDrive/AI/컴퓨터비전


## 👉 4-2. 실습용 데이터 로딩
- 출처(캐글-Food Images) : https://www.kaggle.com/datasets/kmader/food41

- 캐글 이미지 다운로드 받기 (코드 활용)
<br>

    1단계 : 새 노트북 생성 (New notebook)

<img src="https://drive.google.com/uc?export=view&id=1DwjI_St0gywZGj0u46vqE97mEd89tsfO">

    2단계 : 다운로드 코드 작성
<img src="https://drive.google.com/uc?export=view&id=1N6x1o35CDaGluXH6xYs0IOPqI262RANA">

```python
# 경로 변경 (불필요한 경로 제외)
%cd /kaggle/input/food41/images

# 캐글 노트북에서 원하는 디렉토리 선택 다운로드 (압축)
!zip -r /kaggle/working/food.zip cheesecake donuts gyoza nachos takoyaki
```

    3단계 : 다운로드 받은 압축파일 업로드
<img src="https://drive.google.com/uc?export=view&id=1IAxTkF8Q6Cw3ShbxWFR8CKphLwXvRtgQ">

In [ ]:
# 압축풀기
%pwd

!unzip ./data/food.zip -d ./data/food

unzip:  cannot find or open ./data/food.zip, ./data/food.zip.zip or ./data/food.zip.ZIP.


## 👉 4-3. YOLO 폴더 구조 준비
- 훈련용(640장), 평가용(160장), 검증용(200장) (대략적으로 3:1:1)으로 분할
- 폴더 구조 가이드라인 : https://docs.ultralytics.com/ko/datasets/classify/

In [ ]:
# 파일이름, 디렉토리 정보 확인 등 os관련된 명령을 수행하는 라이브러리 불러오기
import os
%pwd

'/content/drive/MyDrive/AI/컴퓨터비전'

In [88]:
# 각 경로에 있는 파일명(리스트) 저장
gyoza_file_names = os.listdir('./food/gyoza')
caesar_salad_file_names = os.listdir('./food/caesar_salad')
bibimbap_file_names = os.listdir('./food/bibimbap')
apple_pie_file_names = os.listdir('./food/apple_pie')
takoyaki_file_names = os.listdir('./food/takoyaki')

In [ ]:
# 파일명 리스트 확인
gyoza_file_names

['1055087.jpg',
 '105877.jpg',
 '1036466.jpg',
 '1020354.jpg',
 '1056855.jpg',
 '1026798.jpg',
 '1047687.jpg',
 '1043205.jpg',
 '1024150.jpg',
 '1060006.jpg',
 '103732.jpg',
 '103201.jpg',
 '1038311.jpg',
 '1055276.jpg',
 '1091447.jpg',
 '1073379.jpg',
 '1009005.jpg',
 '1003912.jpg',
 '2009908.jpg',
 '1978718.jpg',
 '194163.jpg',
 '1965682.jpg',
 '1915725.jpg',
 '1963378.jpg',
 '1948498.jpg',
 '1934382.jpg',
 '1957944.jpg',
 '190757.jpg',
 '1933470.jpg',
 '1905497.jpg',
 '1948493.jpg',
 '1969607.jpg',
 '1946045.jpg',
 '1913551.jpg',
 '1895695.jpg',
 '1866817.jpg',
 '1903351.jpg',
 '1871885.jpg',
 '1879321.jpg',
 '1866658.jpg',
 '1891791.jpg',
 '1879205.jpg',
 '1902549.jpg',
 '1802725.jpg',
 '1881101.jpg',
 '1885087.jpg',
 '1880353.jpg',
 '189419.jpg',
 '1883392.jpg',
 '1891933.jpg',
 '1874905.jpg',
 '1779989.jpg',
 '1734723.jpg',
 '1766220.jpg',
 '175784.jpg',
 '174046.jpg',
 '1732518.jpg',
 '175891.jpg',
 '1721027.jpg',
 '1758252.jpg',
 '174346.jpg',
 '1741953.jpg',
 '1753888.jpg',
 '

In [ ]:
# 각 음식별 파일 갯수 확인
print(len(gyoza_file_names),
      len(bibimbap_file_names),
      len(caesar_salad_file_names),
      len(apple_pie_file_names),
      len(takoyaki_file_names))

1000 1000 1000 1000 1000


In [ ]:
# 폴더생성하기 (data/food_for_yolo, food_for_yolo/train, food_for_yolo/test, food_for_yolo/val), 기존에 만들어진 폴더가 없을 경우에만 생성

#food_for_yolo 폴더 생성
folders = [
    './data/food_for_yolo',
    './data/food_for_train',
    './data/food_for_test',
    './data/food_for_yolo/val'
    ]

if not os.path.exists('./data/food_for_yolo'): # 해당 폴더 없을 때만 생성
  os.makedirs('./data/food_for_yolo') # 생성

if not os.path.exists('./data/food_for_train'): # 해당 폴더 없을 때만 생성
  os.makedirs('./data/food_for_train') # 생성

if not os.path.exists('./data/food_for_test'):
  os.makedirs('./data/food_for_test')
if not os.path.exists('./data/food_for_yolo/val'):
  os.makedirs('./data/food_for_yolo/val')

  for folder in folders:
    if not os.path.exists(folder):
        os.makedirs(folder) # os.mkdir 대신 os.makedirs 사용
        print(f"생성 완료: {folder}")
    else:
        print(f"이미 존재함: {folder}")

In [ ]:
# 하위폴더 생성하기
food_list = ['gyoza', 'bibimbap', 'caesar_salad', 'apple_pie', 'takoyaki']
folder_list = ['./data/food_for_yolo/train', './data/food_for_yolo/test', './data/food_for_yolo/val']

for folder in folder_list:
    for food in food_list:
        path = os.path.join(folder, food) # 경로를 안전하게 합칩니다
        if not os.path.exists(path):
            os.makedirs(path) # 여기를 os.mkdir에서 os.makedirs로 변경
            print(f"생성 완료: {path}")
        else:
            print(f"이미 존재함: {path}")

생성 완료: ./data/food_for_yolo/train/gyoza
생성 완료: ./data/food_for_yolo/train/bibimbap
생성 완료: ./data/food_for_yolo/train/caesar_salad
생성 완료: ./data/food_for_yolo/train/apple_pie
생성 완료: ./data/food_for_yolo/train/takoyaki
생성 완료: ./data/food_for_yolo/test/gyoza
생성 완료: ./data/food_for_yolo/test/bibimbap
생성 완료: ./data/food_for_yolo/test/caesar_salad
생성 완료: ./data/food_for_yolo/test/apple_pie
생성 완료: ./data/food_for_yolo/test/takoyaki
생성 완료: ./data/food_for_yolo/val/gyoza
생성 완료: ./data/food_for_yolo/val/bibimbap
생성 완료: ./data/food_for_yolo/val/caesar_salad
생성 완료: ./data/food_for_yolo/val/apple_pie
생성 완료: ./data/food_for_yolo/val/takoyaki


In [ ]:
# shutil 모듈불러오기
import shutil

In [89]:
# train set 사진 이동하기 (640장, 첫번째 음식)
for img_name in gyoza_file_names[0:640]:
  shutil.move('./food/gyoza/'+img_name,
                './data/food_for_yolo/train/gyoza/'+img_name)

In [95]:
# val set 사진 이동하기 (160장, 첫번째 음식)
for img_name in gyoza_file_names[640:800]:
  shutil.move('./food/gyoza/'+img_name,
                './data/food_for_yolo/train/gyoza/'+img_name)

FileNotFoundError: [Errno 2] No such file or directory: './food/gyoza/2780534.jpg'

In [114]:
# test set 사진 이동하기 (200장, 첫번째 음식)
for img_name in gyoza_file_names[800:]:
   shutil.move('./food/gyoza/'+img_name,
                './data/food_for_yolo/train/gyoza/'+img_name)

FileNotFoundError: [Errno 2] No such file or directory: './food/gyoza/3487080.jpg'

In [96]:
# 나머지 음식 이미지 파일 옮기기
food_file_names = [gyoza_file_names, bibimbap_file_names, caesar_salad_file_names, apple_pie_file_names, takoyaki_file_names]

In [121]:
import shutil
import os

# food_list와 food_file_names_list가 정의되어 있다고 가정합니다.
# zip을 사용하여 음식 종류와 해당 파일 리스트를 하나씩 가져옵니다.
for food, img_list in zip(food_list[1:], food_file_names_list):

    # 1. Train set (0~640)
    for img_name in img_list[0:640]:
        src = f'./food/{food}/{img_name}'
        dst = f'./data/food_for_yolo/train/{food}/{img_name}'
        if os.path.exists(src):
            shutil.move(src, dst)

    # 2. Val set (640~800)
    for img_name in img_list[640:800]:
        src = f'./food/{food}/{img_name}'
        dst = f'./data/food_for_yolo/val/{food}/{img_name}'
        if os.path.exists(src):
            shutil.move(src, dst)

    # 3. Test set (800~)
    for img_name in img_list[800:]:
        src = f'./food/{food}/{img_name}'
        dst = f'./data/food_for_yolo/test/{food}/{img_name}'
        if os.path.exists(src):
            shutil.move(src, dst)

In [120]:
# 1. 파일 이름 리스트들을 먼저 정의해야 합니다 (셀 [88]에서 수행한 작업)
# 만약 이미 정의했다면 이 부분은 건너뛰어도 됩니다.
gyoza_file_names_list = os.listdir('./food/gyoza')
bibimbapa_file_names_list = os.listdir('./food/bibimbap')
caesar_salad_file_names = os.listdir('./food/caesar_salad')
apple_pie_file_names = os.listdir('./food/apple_pie')
takoyaki_file_names = os.listdir('./food/takoyaki')

# 2. 리스트를 하나로 묶어주는 변수 정의 (현재 누락된 부분)
food_file_names_list = [
    gyoza_file_names_list,
    bibimbapa_file_names_list,
    caesar_salad_file_names,
    apple_pie_file_names,
    takoyaki_file_names
]

# 3. food_list도 정의되어 있는지 확인
food_list = ['gyoza', 'bibimbap', 'caesar_salad', 'apple_pie', 'takoyaki']

In [142]:
# 나머지 음식 이미지 파일 옮기기 gyoza, bibimbap, caesar_salad, apple_pie, takoyaki
food_file_names_list = [gyoza_file_names_list, bibimbapa_file_names_list, caesar_salad_file_names, apple_pie_file_names, takoyaki_file_names]

for food, img_list in zip(food_list[1:], food_file_names_list):
  for img_name in img_list[0:640] :
   shutil.move('./food/'+food+'/'+img_name,
                  './food_for_yolo/train/'+food+'/'+img_name)

   for img_name in img_list[640:800]:
    shutil.move('./food/gyoza/'+food+'/'+img_name,
                './food_for_yolo/train/gyoza/'+food+'/'+img_name)

   for img_name in img_list[800:]:
     shutil.move('./food/gyoza/'+food+'/'+img_name,
                './food_for_yolo/train/gyoza/'+food+'/'+img_name)

FileNotFoundError: [Errno 2] No such file or directory: './food/caesar_salad/1025786.jpg'

In [141]:
import shutil
import os

# 1. 대상 폴더들이 제대로 만들어져 있는지 확인
for folder in ['train', 'val', 'test']:
    for food in ['gyoza', 'bibimbap', 'caesar_salad', 'apple_pie', 'takoyaki']:
        path = f'./data/food_for_yolo/{folder}/{food}'
        if not os.path.exists(path):
            os.makedirs(path)

# 2. 파일 이동 수행 (파일이 존재할 때만 이동)
for food, img_list in zip(food_list, food_file_names_list):
    for i, img_name in enumerate(img_list):
        src = os.path.join('./food', food, img_name)

        # 0~640: train, 640~800: val, 800~: test
        if i < 640: folder = 'train'
        elif i < 800: folder = 'val'
        else: folder = 'test'

        dst = os.path.join('./data/food_for_yolo', folder, food, img_name)

        if os.path.exists(src):
            shutil.move(src, dst)
        # 파일이 없으면 이미 옮겨졌으므로 무시함

In [144]:
# 파일 이동여부 확인 (점 '.' 삭제)
train_gyoza = os.listdir('data/food_for_yolo/train/gyoza')
val_gyoza = os.listdir('data/food_for_yolo/val/gyoza')
test_gyoza = os.listdir('data/food_for_yolo/test/gyoza')

# 리스트가 잘 출력되는지 확인
print(f"Train 이미지 수: {len(train_gyoza)}")
print(f"Val 이미지 수: {len(val_gyoza)}")
print(f"Test 이미지 수: {len(test_gyoza)}")

Train 이미지 수: 1000
Val 이미지 수: 0
Test 이미지 수: 0


## 👉 4-4. YOLO 학습하기
-   **YOLO 모델 사용법 가이드라인**: [Ultralytics Usage Modes](https://docs.ultralytics.com/usage/cfg/#modes)

-   **학습 환경 설정**:
    -   CPU를 활용 → 1 epoch만 테스트 수행 권장
    -   GPU를 활용 → 20 epochs 이상 본격적인 학습 수행
-   **학습 모델 정보 (YOLOv11-cls)**:
    -   **공식 문서**: [YOLO 분류(Classification) 모델 가이드](https://docs.ultralytics.com/ko/tasks/classify/)
    -   **사전학습 특징**: 분류 모델은 유일하게 **ImageNet** 데이터셋으로 사전학습되었습니다.
        -   (참고: 객체 탐지, 세그먼트, 포즈 모델 등은 COCO 데이터셋으로 사전학습)

In [111]:
#ultralytics 설치하기
!pip install ultralytics

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.3/41.3 kB 3.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 61.2 MB/s eta 0:00:00


In [145]:
#YOLO 모델 불러오기
from ultralytics import YOLO

In [146]:
# 미리 학습된 모델 가중치 파일을 이용해 YOLO 모델 객체 생성
%pwd

model = YOLO('./data/yolo11n-cls.pt')

In [148]:
# 모델 학습 실행
model.train(
    data='./data/food_for_yolo',  # 점(.) 하나만 사용하거나, 전체 절대 경로를 입력하세요
    epochs = 10, # 학습 횟수
    imgsz = 224 # 224 * 224 크기로 조정하여 학습
)

Ultralytics 8.4.67 🚀 Python-3.12.13 torch-2.11.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=./data/food_for_yolo, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=10, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=224, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=./data/yolo11n-cls.pt, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=train-3, nbs=64, nms=False, opset=None, optimize=False, optimizer=auto, overlap_mask=True, patience=

/usr/local/lib/python3.12/dist-packages/PIL/TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))


train: Scanning /content/drive/MyDrive/AI/컴퓨터비전/data/food_for_yolo/train... 3560 images, 0 corrupt: 100% ━━━━━━━━━━━━ 3560/3560 1.8it/s 32:39
train: New cache created: /content/drive/MyDrive/AI/컴퓨터비전/data/food_for_yolo/train.cache
val: Fast image access ✅ (ping: 1.7±1.8 ms, read: 0.1±0.0 MB/s, size: 38.2 KB)
val: Scanning /content/drive/MyDrive/AI/컴퓨터비전/data/food_for_yolo/val... 640 images, 0 corrupt: 100% ━━━━━━━━━━━━ 640/640 2.0it/s 5:23
val: New cache created: /content/drive/MyDrive/AI/컴퓨터비전/data/food_for_yolo/val.cache
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.001111, momentum=0.9) with parameter groups 39 weight(decay=0.0), 40 weight(decay=0.0005), 40 bias(decay=0.0)
Image sizes 224 train, 224 val
Using 2 dataloader workers
Logging results to /content/drive/MyDrive/AI/컴퓨터비전/runs/classify/train-3
Starting training for 10

ultralytics.utils.metrics.ClassifyMetrics object with attributes:

confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x78551d28d940>
curves: []
curves_results: []
fitness: 0.9789062440395355
keys: ['metrics/accuracy_top1', 'metrics/accuracy_top5']
results_dict: {'metrics/accuracy_top1': 0.957812488079071, 'metrics/accuracy_top5': 1.0, 'fitness': 0.9789062440395355}
save_dir: PosixPath('/content/drive/MyDrive/AI/컴퓨터비전/runs/classify/train-3')
speed: {'preprocess': 0.12901641563587418, 'inference': 1.1868994859440818, 'loss': 0.00036655937378782255, 'postprocess': 0.0004718593771713131}
top1: 0.957812488079071
top5: 1.0

| 출력 항목                                | 설명 |
|-----------------------------------------|------|
| **라이브러리 및 환경 정보**              | `Ultralytics 8.3.82 🚀 Python-3.11.11 torch-2.5.1+cu124 CUDA:0 (Tesla T4, 15095MiB)`<br>- Ultralytics, Python, PyTorch, CUDA 버전 및 사용 GPU(메모리 포함) 정보를 표시 |
| **Trainer 설정 요약**                    | `engine/trainer: task=classify, mode=train, model=./data/yolo11n-cls.pt, data=./data/food_for_yolo/, epochs=20, batch=16, imgsz=224, ... save_dir=runs/classify/train`<br>- 학습 작업(task), 실행 모드(mode), 모델 경로, 데이터셋 경로, 에포크 수, 배치 크기, 이미지 크기 등 주요 하이퍼파라미터 및 설정들을 나열 |
| **데이터셋 스캔 결과 (train/val/test)**    | 예: `train: ... found 3200 images in 5 classes`<br>- 학습(train), 검증(val), 테스트(test) 데이터셋의 파일 경로, 이미지 수, 클래스 수를 확인 |
| **클래스 수 변경**                        | `Overriding model.yaml nc=80 with nc=5`<br>- 원래 모델 설정의 클래스 수 80을 데이터셋에 맞춰 5로 변경 |
| **모델 구성 레이어 테이블**                | 테이블 형태로 각 레이어의 정보를 보여줌 (예: `from`, `n`, `params`, `module`, `arguments`)<br>- 각 레이어가 어떤 모듈로 구성되었는지, 파라미터 수와 인자 값들을 상세히 확인 |
| **모델 요약**                             | `YOLO11n-cls summary: 86 layers, 1,537,509 parameters, 1,537,509 gradients, 3.3 GFLOPs`<br>- 전체 레이어 수, 총 파라미터 및 그래디언트 수, 연산량(GFLOPs)을 요약 |
| **Pretrained 가중치 전이**                 | `Transferred 234/236 items from pretrained weights`<br>- 미리 학습된 가중치에서 몇 개 항목이 성공적으로 전이되었는지를 표시 |
| **TensorBoard 설정 안내**                  | `TensorBoard: Start with 'tensorboard --logdir runs/classify/train'...`<br>- TensorBoard를 통해 학습 과정을 시각화하는 방법을 안내 |
| **AMP 체크 및 파일 다운로드**              | 예: `AMP: running Automatic Mixed Precision (AMP) checks...`,<br>`Downloading ... yolo11n.pt to 'yolo11n.pt'...`,<br>`AMP: checks passed ✅`<br>- 자동 혼합 정밀도(AMP) 체크 실행 및 필요한 파일 다운로드 진행 상황과 체크 성공 여부를 표시 |
| **데이터셋 캐시 생성**                     | 예: `train: New cache created: ...train.cache`,<br>`val: New cache created: ...val.cache`<br>- 학습 및 검증 데이터셋을 스캔한 후 캐시 파일을 생성하여 데이터 로딩 속도를 개선 |
| **Optimizer 설정**                        | 예: `optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937'...`,<br>`optimizer: AdamW(lr=0.001111, momentum=0.9) with parameter groups...`<br>- 옵티마이저 자동 선택 결과와 관련 하이퍼파라미터(learning rate, momentum, weight decay 등)를 출력 |
| **학습 설정 및 시작 안내**                 | `Image sizes 224 train, 224 val`,<br>`Using 2 dataloader workers`,<br>`Logging results to runs/classify/train`,<br>`Starting training for 20 epochs...`<br>- 학습 및 검증 이미지 크기, 데이터 로더 작업자 수, 결과 로그 저장 경로, 총 에포크 수 등 학습 환경을 요약 |
| **Training 진행 상황 출력**               | 예: `Epoch    GPU_mem       loss  Instances       Size` 및 진행률 바<br>- 현재 에포크, GPU 메모리 사용량, 손실 값, 배치 내 이미지(인스턴스) 수, 이미지 크기 등 실시간 학습 진행 상황을 업데이트 |

In [149]:
# 모델검증하기
model.val()

Ultralytics 8.4.67 🚀 Python-3.12.13 torch-2.11.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
YOLO11n-cls summary (fused): 47 layers, 1,532,429 parameters, 0 gradients, 3.2 GFLOPs
train: /content/drive/MyDrive/AI/컴퓨터비전/data/food_for_yolo/train... found 3560 images in 5 classes ✅ 
ERROR ❌ val: /content/drive/MyDrive/AI/컴퓨터비전/data/food_for_yolo/val... found 640 images in 4 classes (requires 5 classes, not 4)
ERROR ❌ test: /content/drive/MyDrive/AI/컴퓨터비전/data/food_for_yolo/test... found 800 images in 4 classes (requires 5 classes, not 4)
val: Fast image access ✅ (ping: 2.2±2.7 ms, read: 11.7±11.9 MB/s, size: 38.2 KB)
val: Scanning /content/drive/MyDrive/AI/컴퓨터비전/data/food_for_yolo/val... 640 images, 0 corrupt: 100% ━━━━━━━━━━━━ 640/640 70.6Mit/s 0.0s
               classes   top1_acc   top5_acc: 100% ━━━━━━━━━━━━ 40/40 6.9it/s 5.8s
                   all      0.958          1
Speed: 0.1ms preprocess, 1.0ms inference, 0.0ms loss, 0.0ms postprocess per image
Results saved t

ultralytics.utils.metrics.ClassifyMetrics object with attributes:

confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x78551b7247d0>
curves: []
curves_results: []
fitness: 0.9789062440395355
keys: ['metrics/accuracy_top1', 'metrics/accuracy_top5']
results_dict: {'metrics/accuracy_top1': 0.957812488079071, 'metrics/accuracy_top5': 1.0, 'fitness': 0.9789062440395355}
save_dir: PosixPath('/content/drive/MyDrive/AI/컴퓨터비전/runs/classify/val')
speed: {'preprocess': 0.12955789845250365, 'inference': 0.9617271203012479, 'loss': 0.000700157821142966, 'postprocess': 0.0035304046889450547}
top1: 0.957812488079071
top5: 1.0

| 출력 항목 | 설명 |
|-----------|------|
| **Ultralytics 8.3.82 🚀 Python-3.11.11 torch-2.5.1+cu124 CUDA:0 (Tesla T4, 15095MiB)** | 사용 중인 Ultralytics 버전, Python, PyTorch, CUDA 정보 및 GPU 모델/메모리 표시 |
| **YOLO11n-cls summary (fused): 47 layers, 1,532,429 parameters, 0 gradients, 3.2 GFLOPs** | 모델 요약 정보: fused(병합)된 상태, 총 47개 레이어, 약 1.53백만 파라미터, 그래디언트 정보(학습 중 업데이트할 값)와 연산량(GFLOPs) 표시 |
| **train: ... found 3200 images in 5 classes ✅** | 학습 데이터셋 경로 및 내용: 3200장의 이미지가 5개 클래스로 구성되어 있음을 의미 |
| **val: ... found 805 images in 5 classes ✅** | 검증 데이터셋 경로 및 내용: 805장의 이미지가 5개 클래스로 구성되어 있음을 의미 |
| **test: ... found 995 images in 5 classes ✅** | 테스트 데이터셋 경로 및 내용: 995장의 이미지가 5개 클래스로 구성되어 있음을 의미 |
| **val: Scanning ... 805 images, 0 corrupt: ...** | 검증 데이터셋 스캔 결과: 805장의 이미지가 문제없이 로드되었고, 손상된 이미지가 없음을 확인 |
| **classes   top1_acc   top5_acc: ...** | 각 클래스별 Top-1, Top-5 정확도 계산 진행 상황을 보여주는 출력 (진행률 바 포함) |
| **all      0.939          1** | 전체 데이터셋에 대한 평가 결과: Top-1 정확도 93.9%, Top-5 정확도 100% |
| **Speed: 0.1ms preprocess, 0.8ms inference, 0.0ms loss, 0.0ms postprocess per image** | 이미지 한 장 처리 시 각 단계(전처리, 추론, 손실 계산, 후처리)에 소요되는 평균 시간 |
| **Results saved to runs/classify/train22** | 평가 결과 및 로그 파일이 저장된 디렉토리 경로 |
| **ultralytics.utils.metrics.ClassifyMetrics object with attributes:** | 평가 결과가 담긴 객체 정보:<br>- **confusion_matrix:** 혼동 행렬 객체<br>- **curves / curves_results:** 성능 곡선 관련 데이터 (여기선 비어 있음)<br>- **fitness:** 전체 평가 점수<br>- **keys, results_dict:** 정확도 관련 메트릭(Top1, Top5)과 fitness 값<br>- **speed:** 각 처리 단계의 속도 정보<br>- **task:** 수행 작업(분류: classify)<br>- **top1, top5:** 최종 Top-1, Top-5 정확도 결과 |

## 👉 4-5. 학습된 모델 활용 예측하기

| 항목 | 설명 |
| --- | --- |
| **예측 결과 요약** | `image 1/1 /content/drive/MyDrive/.../apple_pie/110043.jpg: 224x224 apple_pie 1.00, onion_rings 0.00, donuts 0.00, pizza 0.00, hamburger 0.00, 6.6ms`<br>- 이미지 1장이 처리되었음을 표시 (1/1)<br>- 입력 이미지 크기: 224x224<br>- 각 클래스별 예측 확률: 예를 들어, apple_pie가 100% 확신<br>- 예측 처리 시간: 6.6ms 소요 |
| **처리 속도 정보** | `Speed: 7.5ms preprocess, 6.6ms inference, 0.1ms postprocess per image at shape (1, 3, 224, 224)`<br>- 이미지 전처리(7.5ms), 추론(6.6ms), 후처리(0.1ms)에 걸린 시간<br>- 처리된 이미지의 텐서 형태: (배치 크기 1, 채널 3, 높이 224, 너비 224) |
| **결과 객체 (Results Object)** | `[ultralytics.engine.results.Results object with attributes: ...]`<br>- 예측 결과를 담은 객체로, 주요 속성은 다음과 같습니다: <br>&nbsp;&nbsp;• **boxes**: `None` (객체 감지 작업이 아니므로 바운딩 박스 없음) <br>&nbsp;&nbsp;• **keypoints**: `None` (키포인트 정보 없음) <br>&nbsp;&nbsp;• **masks**: `None` (세그멘테이션 마스크 없음) <br>&nbsp;&nbsp;• **names**: `{0: 'apple_pie', 1: 'donuts', 2: 'hamburger', 3: 'onion_rings', 4: 'pizza'}` (클래스 인덱스와 이름 매핑) <br>&nbsp;&nbsp;• **orig_img**: 원본 이미지 배열 (픽셀 데이터 포함) <br>&nbsp;&nbsp;• **orig_shape**: 원본 이미지의 실제 크기, 예: `(358, 512)` <br>&nbsp;&nbsp;• **path**: 예측에 사용된 이미지 파일의 경로 <br>&nbsp;&nbsp;• **probs**: 예측 확률 정보를 담은 객체 <br>&nbsp;&nbsp;• **save_dir**: 결과가 저장된 디렉토리, 예: `runs/classify/train23` <br>&nbsp;&nbsp;• **speed**: 각 처리 단계(전처리, 추론, 후처리)의 소요 시간 정보를 담은 딕셔너리 |

In [ ]:
# 사진 1장 예측하기


In [154]:
# 여러장 사진 예측하기
results = model('./food_for_yolo/val/gyoza')


FileNotFoundError: ./food_for_yolo/val/gyoza does not exist